### PR TITLE
#22 - check the type of a method reference

### PR DESCRIPTION
**Pull request by csoroiu**
- check the type of the functional interface for lambdas
- check the return type of methods (at the call site)

### CODE REVIEW

**Comment by csoroiu** [comment]
I've redone the tests and rebased the code against master branch.
Also, the file `animal-sniffer-maven-plugin\src\it\jdk14-with-sig14-from-cli\invoker.properties` it is needed because otherwise the `jdk14-with-sig14-from-cli` integration test fails.

Thanks

**Comment by csoroiu** [comment]
on the other hand i reverted `animal-sniffer-maven-plugin\src\it\jdk14-with-sig14-from-cli\invoker.properties` because i just saw that #39 is fixing it more elegantly.

**Comment by rfscholte** [comment]
jdk14-with-sig14-from-cli has been fixed

**Comment by csoroiu** [comment]
given that, i'll rebase again to make sure everything is fine for #22.

thanks

**Comment by rfscholte** [comment]
Travis is still not happy, will investigate it.

**Comment by csoroiu** [comment]
Ah, I just rebased it, i bet it will be happy :).
Travis was happy before,  in build https://travis-ci.org/mojohaus/animal-sniffer/builds/261932108.
Build just started and it will take like 10-15 minutes to complete.

**Comment by csoroiu** [comment]
actually i take that back, apparently it does not like the test for #32 👎

**Comment by csoroiu** [comment]
build passed after a second rebase. 
very interesting anyway.
when I ran the test on my machine before debug got enabled, it passed.

File: animal-sniffer/src/main/java/org/codehaus/mojo/animal_sniffer/SignatureChecker.java [hunk 1]
```diff
@@ -370,16 +372,69 @@ public AnnotationVisitor visitAnnotation( String annoDesc, boolean visible )
                     return super.visitAnnotation( annoDesc, visible );
                 }
 
-                
+                private static final String LAMBDA_METAFACTORY = "java/lang/invoke/LambdaMetafactory";
+
+                @Override
+                public void visitInvokeDynamicInsn( String name, String desc, Handle bsm, Object... bsmArgs )
+                {
+                    if ( bsm.getOwner().equals( LAMBDA_METAFACTORY ) )
```
Comment (rfscholte) [inline animal-sniffer/src/main/java/org/codehaus/mojo/animal_sniffer/SignatureChecker.java]: Please change style to CONSTANT.equals(variable) to reduce change of NPE.

File: animal-sniffer/src/main/java/org/codehaus/mojo/animal_sniffer/SignatureChecker.java [hunk 2]
```diff
@@ -370,16 +372,69 @@ public AnnotationVisitor visitAnnotation( String annoDesc, boolean visible )
...
+                    {
+                        if ( bsm.getName().equals( "metafactory" ) ||
```
Comment (rfscholte) [inline animal-sniffer/src/main/java/org/codehaus/mojo/animal_sniffer/SignatureChecker.java]: Also switch equal statements here

Comment (csoroiu) [inline animal-sniffer/src/main/java/org/codehaus/mojo/animal_sniffer/SignatureChecker.java]: Just did a force push

Thanks

**Comment by csoroiu** [comment]
for some reason all jdk7 builds failed.

the issue seems to be related to he maven cache
```
fetching master/cache--jdk-oraclejdk7.tgz
could not download cache
```
not related to my commit.

**Comment by rfscholte** [comment]
Those jobs were ignored, the red sign is a bit distracting

**Comment by csoroiu** [comment]
thanks

**Pull request merged**